src/vmm/src/builder.rs

```diff
@@ -427,6 +429,36 @@ pub fn build_microvm_from_snapshot(
         vcpu_count,
     )?;
 
+    #[cfg(target_arch = "x86_64")]
+    // Check if we need to scale the TSC.
+    // We start by checking if the CPU model in the snapshot is
+    // the same as this host's. If they are the same, we don't
+    // need to do anything else.
+    if !is_same_model(&microvm_state.vcpu_states[0].cpuid) {
+        // Extract the TSC freq from the state.
+        // No TSC freq in snapshot means we have to fail-fast.
+        let state_tsc = microvm_state.vcpu_states[0]
+            .tsc_khz
+            .ok_or_else(|| {
+                MicrovmStateError::IncompatibleState(
+                    "Error configuring the TSC, frequency not \
+                    present in snapshot."
+                        .to_string(),
+                )
+            })
+            .map_err(RestoreMicrovmState)?;
+
+        // Scale the TSC frequency for all VCPUs, if needed.
+        if vcpus[0].kvm_vcpu.is_tsc_scaling_required(state_tsc) {
+            for vcpu in &vcpus {
+                vcpu.kvm_vcpu
+                    .set_tsc_khz(state_tsc)
+                    .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
+                    .map_err(RestoreMicrovmState)?;
+            }
+        }
+    }
```

Comment:
I would move this code to the beginning of `restore_vcpu_states()`

Reply:
When Firecracker is used correctly, we shouldn't encounter issues related to setting the TSC. Failing should be more of an edge case. So I think that it's worth the move for readability.

Reply:
The problem is that by the time `restore_vcpu_states` is called, the VCPUs are already moved to their own threads and we run into the same problem as before regarding the redundant checks. Furthermore, it would just add extra complexity to the code to try and send these check specific fields through the `VcpuHandle` interface. I lean towards keeping it as is.